src/xmerl_c14n.erl

```diff
@@ -0,0 +1,527 @@
+    case Exp of
+        {Ns, Nme} -> canon_name(Ns, Nme, Nsp);
+        _ -> canon_name([], Name, Nsp)
+    end.
+
+%% @doc Compares two XML attributes for c14n purposes
+-spec attr_lte(A :: #xmlAttribute{}, B :: #xmlAttribute{}) -> true | false.
+attr_lte(AttrA, AttrB) ->
+    logger:debug("AttrA: ~p AttrB: ~p~n", [AttrA, AttrB]),
+    A = canon_name(AttrA), B = canon_name(AttrB),
+    PrefixedA = case AttrA#xmlAttribute.nsinfo of {_, _} -> true; _ -> false end,
+    PrefixedB = case AttrB#xmlAttribute.nsinfo of {_, _} -> true; _ -> false end,
+    if (PrefixedA) andalso (not PrefixedB) ->
+        false;
+    (not PrefixedA) andalso (PrefixedB) ->
+        true;
+    true ->
+        A =< B
+    end.
+
+%% @doc Cleans out all namespace definitions from an attribute list and returns it sorted.
+%% @private
+-spec clean_sort_attrs(Attrs :: [#xmlAttribute{}]) -> [#xmlAttribute{}].
+clean_sort_attrs(Attrs) ->
+    lists:sort(fun(A,B) ->
+        attr_lte(A, B)
+    end, lists:filter(fun(Attr) ->
+        case Attr#xmlAttribute.nsinfo of
+            {"xmlns", _} -> false;
+            _ -> case Attr#xmlAttribute.name of
+                'xmlns' -> false;
+                _ -> true
+            end
+        end
+    end, Attrs)).
+
+%% @doc Returns the list of namespace prefixes "needed" by an element in canonical form
+%% @private
+-spec needed_ns(Elem :: #xmlElement{}, InclNs :: [string()]) -> [string()].
+needed_ns(#xmlElement{nsinfo = NsInfo, attributes = Attrs}, InclNs) ->
+    NeededNs1 = case NsInfo of
+        {Nas, _} -> [Nas];
+        _ -> []
+    end,
+    %% show through namespaces that apply at the bottom level? this part of the spec is retarded
+    %%KidElems = [K || K <- Kids, element(1, K) =:= xmlElement],
+    NeededNs2 = NeededNs1, %case KidElems of
+        %%[] -> [K || {K,V} <- E#xmlElement.namespace#xmlNamespace.nodes];
+        %%_ -> NeededNs1
+    %%end,
+    lists:foldl(fun(Attr, Needed) ->
+        case Attr#xmlAttribute.nsinfo of
+            {"xmlns", Prefix} ->
+                case lists:member(Prefix, InclNs) and not lists:member(Prefix, Needed) of
+                    true -> [Prefix | Needed];
+                    _ -> Needed
+                end;
+            {Ns, _Name} ->
+                case lists:member(Ns, Needed) of
+                    true -> Needed;
+                    _ -> [Ns | Needed]
+                end;
+            _ -> Needed
+        end
+    end, NeededNs2, Attrs).
+
+%% @doc Make xml ok to eat, in a non-quoted situation.
+%% @private
+-spec xml_safe_string(term()) -> string().
+xml_safe_string(Term) -> xml_safe_string(Term, false).
+
+%% @doc Make xml ok to eat
+%% @private
+-spec xml_safe_string(String :: term(), Quotes :: boolean()) -> string().
+xml_safe_string(Atom, Quotes) when is_atom(Atom) -> xml_safe_string(atom_to_list(Atom), Quotes);
+xml_safe_string(Bin, Quotes) when is_binary(Bin) -> xml_safe_string(binary_to_list(Bin), Quotes);
```

Comment:
How does this handle UTF-8?

Reply:
It will correctly handle utf8 with binaries, but not atoms.

```
19> binary_to_list(<<"föö"/utf8>>).
"fÃ¶Ã¶"
20> atom_to_list(föö).
"föö"
```

This also the original authors code, and I don't think we use atoms in `#xmlElements` anywere in ews, so it should be fine. There is an utf-8 tests at the bottom that seems to work fine.